### PR TITLE
add sts e2e case for pr

### PR DESCRIPTION
**Pull request by changluyi**
https://github.com/kubeovn/kube-ovn/pull/3777

# Pull Request

- [ ] Make sure you have followed [Kube-OVN Code Style](https://github.com/kubeovn/kube-ovn/blob/master/CODE_STYLE.md).

## What type of this PR

Examples of user facing changes:

- Features
- Bug fixes
- Docs
- Tests

## Which issue(s) this PR fixes

Fixes #(issue-number)

### CODE REVIEW

File: test/e2e/framework/statefulset.go [hunk 1]
```diff
@@ -96,6 +104,62 @@ func (c *StatefulSetClient) WaitToDisappear(name string, _, timeout time.Duratio
...
+	WaitUntil(2*time.Second, timeout, func(_ context.Context) (bool, error) {
+		var err error
+		sts = c.Get(name)
+		unstructured := &unstructured.Unstructured{}
+		if unstructured.Object, err = runtime.DefaultUnstructuredConverter.ToUnstructured(sts); err != nil {
+			return false, err
+		}
+
+		dsv := &polymorphichelpers.StatefulSetStatusViewer{}
+		msg, done, err := dsv.Status(unstructured, 0)
```
Comment (zhangzujian) [inline test/e2e/framework/statefulset.go:148]:
```suggestion
		viewer := &polymorphichelpers.StatefulSetStatusViewer{}
		msg, done, err := viewer.Status(unstructured, 0)
```

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+		ginkgo.By("Creating statefulset " + stsName)
+		sts := framework.MakeStatefulSet(stsName, stsName, int32(replicas), labels, framework.PauseImage)
+		sts = stsClient.CreateSync(sts)
+		ginkgo.By("Delete pod for statefulset " + stsName)
+		pod2Name := stsName + "-2"
+		pod2 := podClient.GetPod(pod2Name)
+		pod2IP := pod2.Annotations[util.IPAddressAnnotation]
+		err := podClient.Delete(pod2Name)
+		framework.ExpectNoError(err, "failed to delete pod "+pod2Name)
+		stsClient.WaitForRunningAndReady(sts)
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:49]:
```suggestion
		ginkgo.By("Waiting for statefulset " + stsName + " to be ready")
		stsClient.WaitForRunningAndReady(sts)
```

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+		pod2 = podClient.GetPod(pod2Name)
+		framework.ExpectEqual(pod2.Annotations[util.IPAddressAnnotation], pod2IP)
+
+		ginkgo.By("Scale sts replica to 1 and Scale replica to 3 " + stsName)
+		patchSts := framework.MakeStatefulSet(stsName, stsName, int32(1), labels, framework.PauseImage)
+		stsClient.PatchSync(sts, patchSts)
+		patchSts = framework.MakeStatefulSet(stsName, stsName, int32(3), labels, framework.PauseImage)
+		sts = stsClient.Get(stsName)
+		stsClient.PatchSync(sts, patchSts)
+		stsClient.WaitForRunningAndReady(patchSts)
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:59]:
```suggestion
		ginkgo.By("Waiting for statefulset " + stsName + " to be ready")
		stsClient.WaitForRunningAndReady(patchSts)
```

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+)
+
+var _ = framework.Describe("[group:pod]", func() {
+	f := framework.NewDefaultFramework("pod")
+
+	var cs clientset.Interface
+	var podClient *framework.PodClient
+	var stsClient *framework.StatefulSetClient
+	var stsName string
+	var image string
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:19]:
```suggestion
```

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+		sts = stsClient.CreateSync(sts)
+		ginkgo.By("Delete pod for statefulset " + stsName)
+		pod2Name := stsName + "-2"
+		pod2 := podClient.GetPod(pod2Name)
+		pod2IP := pod2.Annotations[util.IPAddressAnnotation]
+		err := podClient.Delete(pod2Name)
+		framework.ExpectNoError(err, "failed to delete pod "+pod2Name)
+		stsClient.WaitForRunningAndReady(sts)
+		pod2 = podClient.GetPod(pod2Name)
+		framework.ExpectEqual(pod2.Annotations[util.IPAddressAnnotation], pod2IP)
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:51]: This case has been inclued in `test/e2e/kube-ovn/ipam/ipam.go`, and there is no need to test it twice.

Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go:43]: fixed

Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go:43]: this is the necessary step for repeat bug

File: test/e2e/kube-ovn/pod/statefulset.go (hunk 4 above)
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+		ginkgo.By("Scale sts replica to 1 and Scale replica to 3 " + stsName)
+		patchSts := framework.MakeStatefulSet(stsName, stsName, int32(1), labels, framework.PauseImage)
+		stsClient.PatchSync(sts, patchSts)
```
[the remark this answers is not in the document]
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go:47]: fixed

File: test/e2e/framework/statefulset.go (hunk 1 above)
Comment (changluyi) [reply test/e2e/framework/statefulset.go]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go (hunk 3 above)
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go:55]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+		pod2Name := stsName + "-2"
+		pod2 := podClient.GetPod(pod2Name)
+		pod2IP := pod2.Annotations[util.IPAddressAnnotation]
+		err := podClient.Delete(pod2Name)
+		framework.ExpectNoError(err, "failed to delete pod "+pod2Name)
+		stsClient.WaitForRunningAndReady(sts)
+		pod2 = podClient.GetPod(pod2Name)
+		framework.ExpectEqual(pod2.Annotations[util.IPAddressAnnotation], pod2IP)
+
+		ginkgo.By("Scale sts replica to 1 and Scale replica to 3 " + stsName)
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:45]:
```suggestion
		ginkgo.By("Scale sts replicas to 1)
```

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+		patchSts := framework.MakeStatefulSet(stsName, stsName, int32(1), labels, framework.PauseImage)
+		stsClient.PatchSync(sts, patchSts)
+		ginkgo.By("Waiting for statefulset " + stsName + " to be ready")
+		stsClient.WaitForRunningAndReady(patchSts)
+
+		patchSts = framework.MakeStatefulSet(stsName, stsName, int32(3), labels, framework.PauseImage)
```
Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:51]:
```suggestion
		ginkgo.By("Scale sts replicas to 3")
		patchSts = framework.MakeStatefulSet(stsName, stsName, int32(3), labels, framework.PauseImage)
```

Comment (zhangzujian) [inline test/e2e/kube-ovn/pod/statefulset.go:51]:
```suggestion
		sts = stsClient.Get(stsName)
		patchSts = sts.DeepCopy()
		*patchSts.Spec.Replicas = 3
```

Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+		pod2 := podClient.GetPod(pod2Name)
+		pod2IP := pod2.Annotations[util.IPAddressAnnotation]
+		err := podClient.Delete(pod2Name)
+		framework.ExpectNoError(err, "failed to delete pod "+pod2Name)
+		stsClient.WaitForRunningAndReady(sts)
+		pod2 = podClient.GetPod(pod2Name)
+		framework.ExpectEqual(pod2.Annotations[util.IPAddressAnnotation], pod2IP)
+
+		ginkgo.By("Scale sts replica to 1 and Scale replica to 3 " + stsName)
+		patchSts := framework.MakeStatefulSet(stsName, stsName, int32(1), labels, framework.PauseImage)
```
[the remark this answers is not in the document]
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go (hunk 8 above)
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go]: fixed

File: test/e2e/kube-ovn/pod/statefulset.go (hunk 7 above)
Comment (changluyi) [reply test/e2e/kube-ovn/pod/statefulset.go]: fixed

**Review by zhangzujian — APPROVED** [review]

**Pull request merged**